crates/nu-plugin/src/util/with_custom_values_in.rs

```diff
@@ -0,0 +1,102 @@
+use nu_protocol::{CustomValue, IntoSpanned, ShellError, Spanned, Value};
+
+/// Do something with all [`CustomValue`]s recursively within a `Value`. This is not limited to
+/// plugin custom values.
+///
+/// `LazyRecord`s will be collected to plain values for completeness.
+pub fn with_custom_values_in<E>(
+    value: &mut Value,
+    mut f: impl FnMut(Spanned<&mut (dyn CustomValue + '_)>) -> Result<(), E>,
+) -> Result<(), E>
+where
+    E: From<ShellError>,
+{
+    value.recurse_mut(&mut |value| {
+        let span = value.span();
+        match value {
+            Value::Custom { val, .. } => {
+                // Operate on a CustomValue.
+                f(val.as_mut().into_spanned(span))
+            }
+            // LazyRecord would be a problem for us, since it could return something else the
+            // next time, and we have to collect it anyway to serialize it. Collect it in place,
+            // and then use the result
+            Value::LazyRecord { val, .. } => {
+                *value = val.collect()?;
+                Ok(())
+            }
+            _ => Ok(()),
```

Comment:
Is this code only executed when crossing the engine/plugin boundary?

Reply:
Yes, it's mostly called by various functions that prepare data to cross the boundary.